### PR TITLE
Switch statefulset controller to shared informers

### PR DESCRIPTION
**Pull request by ncdc**
Originally part of #40097 

I *think* the controller currently makes a deep copy of a StatefulSet before it mutates it, but I'm not 100% sure. For those who are most familiar with this code, could you please confirm?

@beeps @smarterclayton @ingvagabund @sttts @liggitt @deads2k @kubernetes/sig-apps-pr-reviews @kubernetes/sig-scalability-pr-reviews @timothysc @gmarek @wojtek-t

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set.go [hunk 1]
```diff
@@ -114,28 +110,33 @@ func NewStatefulSetController(podInformer cache.SharedIndexInformer, kubeClient
 			},
 			DeleteFunc: ssc.enqueueStatefulSet,
 		},
+		statefulSetResyncPeriod,
```
Comment (kargakis) [inline pkg/controller/statefulset/stateful_set.go]: We should be able to switch to the default resync period but probably not before adding a pvc cache ... Add a TODO

Comment (ncdc) [inline pkg/controller/statefulset/stateful_set.go]: @kargakis do you want me to plumb in using a PVC informer now?

**Comment by ncdc** [comment]
@k8s-bot kubemark e2e test this, infra flake

```
W0215 08:25:47.304] ERROR: (gcloud.components.install) The component [alpha] failed to download.
W0215 08:25:47.304] 
W0215 08:25:47.304] ('The read operation timed out',)
```

File: pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/statefulset/stateful_set.go]: Not necessary I think but an issue to extend the resync period and also add the PVC informer would be nice.

**Comment by smarterclayton** [comment]
@foxish @kow3ns

**Comment by kow3ns** [comment]
@ncdc The code only mutates the StatefulSet.Status.Replicas field. The relevant code is [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_pod_control.go#L139-L165). The mutation is performed on a deep copy. The supplied *apps.StatefulSet parameter is treated as an inout parameter, and it is mutated, but it is only mutated when it successfully commits its current status API server. The value of the deep copy is re-assigned to the input parameter.

**Comment by kow3ns** [comment]
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/controller/statefulset/stateful_set.go, line 34 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-Kd2FrjQqwNXyR6GZUGG:-Kd2FrjQqwNXyR6GZUGH:bs9ai77) ([raw file](https://github.com/kubernetes/kubernetes/blob/a844b3b7c851efad02c95a5977155a0c35af7490/pkg/controller/statefulset/stateful_set.go#L34)):*
> ```Go
> 	"k8s.io/client-go/tools/record"
> 	"k8s.io/client-go/util/workqueue"
> 	"k8s.io/kubernetes/pkg/api"
> ```

space between client-go and kubernetes?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by smarterclayton** [comment]
We're still combining everything under `k8s.io` into one import block.

**Comment by ncdc** [comment]
> The code only mutates the StatefulSet.Status.Replicas field. The relevant code is here. The mutation is performed on a deep copy. The supplied *apps.StatefulSet parameter is treated as an inout parameter, and it is mutated, but it is only mutated when it successfully commits its current status API server. The value of the deep copy is re-assigned to the input parameter.

@kow3ns ok, thanks. So I'll need to fix this in my PR, since the reassignment of the parameter results in changing the data in the shared cache.

**Comment by spiffxp** [comment]
FWIW I'm probably not the right person to be assigned this PR, I'm not in any of the OWNERS files.  cc'ing the approvers listed at cmd/kube-controller-manager/OWNERS

/cc @deads2k @lavalamp @mikedanese

**Comment by ncdc** [comment]
I've added a new commit that stops mutating the stateful sets and pods retrieved from the shared informer caches. PTAL.

**Comment by ncdc** [comment]
@k8s-bot cvm gke e2e test this

File: pkg/controller/statefulset/stateful_set_control_test.go [hunk 2]
```diff
@@ -726,6 +915,11 @@ func scaleDownStatefulSetControl(set *apps.StatefulSet, ssc StatefulSetControlIn
 			spc.podsIndexer.Delete(pods[ordinal])
 		}
 		if err := ssc.UpdateStatefulSet(set, pods); err != nil {
+			fmt.Printf("ANDY3: %v\n", err)
```
Comment (smarterclayton) [inline pkg/controller/statefulset/stateful_set_control_test.go]: debug output

**Comment by ncdc** [comment]
Rebased, squashed, debugging removed

**Comment by kow3ns** [comment]
Reviewed 2 of 4 files at r2, 3 of 4 files at r3.
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/controller/statefulset/stateful_pod_control.go, line 38 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBdB_tXGOuaw2tXT2b:-KdBdB_tXGOuaw2tXT2c:b-1l6oj4) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_pod_control.go#L38)):*
> ```Go
> // implementation provides for PVC creation, ordered Pod creation, ordered Pod termination, and Pod identity enforcement.
> // Like controller.PodControlInterface, it is implemented as an interface to provide for testing fakes.
> type StatefulPodControlInterface interface {
> ```

Update comments to address changes made to the interface contract with respect to the mutation of the parameter.

---

*[pkg/controller/statefulset/stateful_pod_control.go, line 91 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBdbwS_pf13fLNc6lc:-KdBdbwS_pf13fLNc6ld:b-xef5yy) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_pod_control.go#L91)):*
> ```Go
> 	podCopy := obj.(*v1.Pod)
> 	attemptedUpdate := false
> 	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
> ```

Does this ever stop retrying. We will now have a backoff in the workqueue and here, what are the implications?

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdBflEHersRYhTxE9f9:b-jh0r5p) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
> ```Go
> 	// if the current number of replicas has changed update the statefulSets replicas
> 	if set.Status.Replicas != int32(ready) {
> 		obj, err := api.Scheme.Copy(set)
> ```

Move the StatefulSet deep copy into UpdateStatefulSetReplicas or move the pod deep copy out of UpdateStatefulPod for consistency.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/controller/statefulset/stateful_pod_control.go, line 38 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBdB_tXGOuaw2tXT2b:-KdBqWWh8uqAt-lC69Go:bebrt1w) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_pod_control.go#L38)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Update comments to address changes made to the interface contract with respect to the mutation of the parameter.
</blockquote></details>

Will do

---

*[pkg/controller/statefulset/stateful_pod_control.go, line 91 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBdbwS_pf13fLNc6lc:-KdBqa5D7t0TOud_NC2G:b-jmepmh) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_pod_control.go#L91)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Does this ever stop retrying. We will now have a backoff in the workqueue and here, what are the implications?
</blockquote></details>

It does stop retrying, yes. You'll see that the unit test for conflict failure always returns a conflict.

I would think having backoff here would be preferable to not having it. Take a look at the parameters for `retry.DefaultBackoff`.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdBrDu6VXiE9hOzpvHl:b32w1t) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Move the StatefulSet deep copy into UpdateStatefulSetReplicas or move the pod deep copy out of UpdateStatefulPod for consistency.
</blockquote></details>

Ok

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdBs-JdbLQfxBuo_JDM:bqh6vc2) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, ncdc (Andy Goldstein) wrote…</i></summary><blockquote>

Ok
</blockquote></details>

Actually, I did it this way for a reason. I need to mutate the StatefulSet before I pass it into UpdateStatefulSetReplicas because I want to use the mutated StatefulSet afterwards. I left the pod mutation inside because I don't need access to the updated pod.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
@kow3ns updated, ptal.

**Comment by kow3ns** [comment]
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/controller/statefulset/stateful_pod_control.go, line 91 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBdbwS_pf13fLNc6lc:-KdC7lg-uA3Vxrk7UPu6:bv8diqd) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_pod_control.go#L91)):*
<details><summary><i>Previously, ncdc (Andy Goldstein) wrote…</i></summary><blockquote>

It does stop retrying, yes. You'll see that the unit test for conflict failure always returns a conflict.

I would think having backoff here would be preferable to not having it. Take a look at the parameters for `retry.DefaultBackoff`.
</blockquote></details>

Oh definitely, having a library method that uses a backoff and is exercised in other controllers is the right approach.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdC8aI7xODDSzJHs3T6:bcvtzp1) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, ncdc (Andy Goldstein) wrote…</i></summary><blockquote>

Actually, I did it this way for a reason. I need to mutate the StatefulSet before I pass it into UpdateStatefulSetReplicas because I want to use the mutated StatefulSet afterwards. I left the pod mutation inside because I don't need access to the updated pod.
</blockquote></details>

Could we update both methods to the following signatures. It will require a few updates to the stateful_pod_control_test.go but would provide uniform semantics for the interface. All parameters will be in parameters and all mutations will be signaled by return values.

```golang
        // UpdateStatefulPod Updates a Pod in a StatefulSet. If the Pod already has the correct identity and stable
	// storage this method is a no-op. If the Pod must be mutated to conform to the Set, it is mutated and updated.
	// pod is not mutated. If the returned error is nil, the returned Pod is not nil, and it contains the updated value of pod.
	UpdateStatefulPod(set *apps.StatefulSet, pod *v1.Pod) (*v1.Pod, error)
	
	// UpdateStatefulSetStatus Updates the Status.Replicas of set to replicas. set is not mutated. If the returned error is nil, the 
        // returned StatefulSet is not nil, and it contains the updated value of set. 
	UpdateStatefulSetReplicas(set *apps.StatefulSet, replicas int32) (*apps.StatefulSet, error)
```

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
Review status: 10 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdCChNoeYZFkBrgQeQ-:b-7lnv6l) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Could we update both methods to the following signatures. It will require a few updates to the stateful_pod_control_test.go but would provide uniform semantics for the interface. All parameters will be in parameters and all mutations will be signaled by return values.

```golang
        // UpdateStatefulPod Updates a Pod in a StatefulSet. If the Pod already has the correct identity and stable
	// storage this method is a no-op. If the Pod must be mutated to conform to the Set, it is mutated and updated.
	// pod is not mutated. If the returned error is nil, the returned Pod is not nil, and it contains the updated value of pod.
	UpdateStatefulPod(set *apps.StatefulSet, pod *v1.Pod) (*v1.Pod, error)
	
	// UpdateStatefulSetStatus Updates the Status.Replicas of set to replicas. set is not mutated. If the returned error is nil, the 
        // returned StatefulSet is not nil, and it contains the updated value of set. 
	UpdateStatefulSetReplicas(set *apps.StatefulSet, replicas int32) (*apps.StatefulSet, error)
```

</blockquote></details>

I'm ok doing that if that's what you want. I had actually started down that road but it isn't really consistent with the other controllers so I opted for this way instead.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by kow3ns** [comment]
*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdCHG56IiDB5c87SYhS:b-28r7yw) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, ncdc (Andy Goldstein) wrote…</i></summary><blockquote>

I'm ok doing that if that's what you want. I had actually started down that road but it isn't really consistent with the other controllers so I opted for this way instead.
</blockquote></details>

 I'd also be okay with performing the pod copies outside of the interface, provided that its consistent for both. Unless you have a strong reason for having the UpdateStatefulPod treat the parameter as immutable and UpdateStatefulSetReplicas treat the parameter as mutable.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
Review status: 10 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdCJy-w6LwAYjqLwgYl:bp7ba8l) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

 I'd also be okay with performing the pod copies outside of the interface, provided that its consistent for both. Unless you have a strong reason for having the UpdateStatefulPod treat the parameter as immutable and UpdateStatefulSetReplicas treat the parameter as mutable.
</blockquote></details>

I only want to perform a deep copy if it's necessary or more likely than not. It didn't appear that the pod mutation was needed outside of UpdateStatefulPod, so I left the copy inside. I'd be happy to move the copy outside, along with the check to see if it needs a change or not. Would that work for you?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by kow3ns** [comment]
*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdCKwhSrjqxzwrFuHSH:b2nhswy) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, ncdc (Andy Goldstein) wrote…</i></summary><blockquote>

I only want to perform a deep copy if it's necessary or more likely than not. It didn't appear that the pod mutation was needed outside of UpdateStatefulPod, so I left the copy inside. I'd be happy to move the copy outside, along with the check to see if it needs a change or not. Would that work for you?
</blockquote></details>

Yes

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by ncdc** [comment]
Review status: 8 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/controller/statefulset/stateful_set_control.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482#-KdBflEHersRYhTxE9f8:-KdCgC-qxTjF-SNrNTNx:b-kjc5ac) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6dbfbdefb1708f6f0f876b2d3abcf4652167ace/pkg/controller/statefulset/stateful_set_control.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Yes
</blockquote></details>

@kow3ns ok, I've updated it. The test `TestDefaultStatefulSetControlUpdatePodFailure` turned out to a "fun" one to fix. PTAL.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by kow3ns** [comment]
Reviewed 4 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by kow3ns** [comment]
@k8s-bot verify test this

**Comment by ncdc** [comment]
@kow3ns lmk if it looks ok and I'll squash when I'm back at my desk.

**Comment by ncdc** [comment]
Rebased, squashed, and hopefully fixed the vet issue

**Comment by kow3ns** [comment]
@k8s-bot e2e test this

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 3]
```diff
@@ -137,31 +133,21 @@ func (spc *realStatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod
...
+		if err == nil {
+			return nil
 		}
-		conflicting, err := spc.client.Apps().StatefulSets(setCopy.Namespace).Get(setCopy.Name, metav1.GetOptions{})
-		if err != nil {
-			return err
+
+		updateErr := err
+
+		if updated, err := spc.client.Apps().StatefulSets(set.Namespace).Get(set.Name, metav1.GetOptions{}); err == nil {
```
Comment (kargakis) [inline pkg/controller/statefulset/stateful_pod_control.go]: Shouldn't this be a cache lookup?

Comment (ncdc) [inline pkg/controller/statefulset/stateful_pod_control.go]: Fixing it now.

**Comment by ncdc** [comment]
Updated to look up the stateful set from the cache in the event of a failed update of the set's replica count.

**Comment by kargakis** [comment]
> Updated to look up the stateful set from the cache in the event of a failed update of the set's replica count.

You'll want to deep-copy that or you'll make @deads2k sad. Also there is a pod read call in the other retry func that can be turned into a cache lookup. I just don't know if we want to go all-in with our caches... :)

**Comment by ncdc** [comment]
> You'll want to deep-copy that or you'll make @deads2k sad

And me too :-(. Will fix.

> Also there is a pod read call in the other retry func that can be turned into a cache lookup. I just don't know if we want to go all-in with our caches... :)

I'll update that too. Thanks.

**Comment by ncdc** [comment]
@kargakis updated

**Comment by ncdc** [comment]
@kow3ns please review the latest update when you get a chance. Thanks!

**Comment by ncdc** [comment]
@k8s-bot cvm gce e2e test this

**Comment by kow3ns** [comment]
Reviewed 1 of 1 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41482)*

**Comment by smarterclayton** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this

**Comment by ncdc** [comment]
@k8s-bot cvm gce e2e test this

**Comment by ncdc** [comment]
Anyone have any other comments?

**Comment by kow3ns** [comment]
@k8s-bot non-cri e2e test this

**Comment by ncdc** [comment]
The non-cri e2e failure looks like the recently-created https://github.com/kubernetes/kubernetes/issues/41889. I don't think it's unique to this PR.

**Comment by deads2k** [comment]
lgtm

**Comment by ncdc** [comment]
Green tests 😄

**Pull request merged**